test/functional/legacy/011_spec.lua

```diff
@@ -0,0 +1,231 @@
+-- Tests for autocommands
+-- - FileWritePre		writing a compressed file
+-- - FileReadPost		reading a compressed file
+-- - BufNewFile			reading a file template
+-- - BufReadPre			decompressing the file to be read
+-- - FilterReadPre		substituting characters in the temp file
+-- - FilterReadPost		substituting characters after filtering
+-- - FileReadPre		set options for decompression
+-- - FileReadPost		decompress the file
+-- Note: This test is skipped if "gzip" is not available.
+-- $GZIP is made empty, "-v" would cause trouble.
+-- Use a FileChangedShell autocommand to avoid a prompt for "Xtestfile.gz"
+-- being modified outside of Vim (noticed on Solaris).
+
+local helpers, lfs = require('test.functional.helpers'), require('lfs')
+local clear, execute, expect, eq, neq, dedent, write_file =
+  helpers.clear, helpers.execute, helpers.expect, helpers.eq, helpers.neq,
+  helpers.dedent, helpers.write_file
+local wait = helpers.wait -- only for debugging
+
+local function has_gzip()
+  return os.execute('gzip --help >/dev/null 2>&1') == 0
+end
+
+describe('file reading, writing and bufnew and filter autocommands', function()
+  local text1 = dedent([[
+      start of testfile
+      line 2	Abcdefghijklmnopqrstuvwxyz
+      line 3	xxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
+      line 4	Abcdefghijklmnopqrstuvwxyz
+      line 5	xxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
+      line 6	Abcdefghijklmnopqrstuvwxyz
+      line 7	xxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
+      line 8	Abcdefghijklmnopqrstuvwxyz
+      line 9	xxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
+      line 10 Abcdefghijklmnopqrstuvwxyz
+      end of testfile]])
+  setup(function()
+    write_file('Xtest.c', [[
+      /*
+       * Here is a new .c file
+       */
+      ]])
+  end)
+  before_each(function()
+    clear()
+    if has_gzip() then
+      write_file('Xtestfile', text1..'\n')
+      -- First we need to compress the test file with gzip.
+      os.execute('gzip --force Xtestfile')
+      -- This should create Xtestfile.gz and delete Xtestfile.
+      neq(nil, lfs.attributes('Xtestfile.gz'))
+      eq(nil, lfs.attributes('Xtestfile'))
+    end
+  end)
+  teardown(function()
+    os.remove('Xtestfile.gz')
+    os.remove('Xtest.c')
+--    os.remove('foo.c')
+    os.remove('test.out')
+  end)
+
+  if not has_gzip() then
+    pending('skipped (missing `gzip` utility)', function() end)
+  else
+
+    it('FileReadPost (using gzip)', function()
+      execute('let $GZIP = ""')
+      --execute('au FileChangedShell * echo "caught FileChangedShell"')
+      execute('set bin')
+      execute("au FileReadPost    *.gz   '[,']!gzip -d")
+      -- Read and decompress the testfile.
+      execute('$r Xtestfile.gz')
+      wait()
+      expect('\n'..text1)
+    end)
+
+    it('BufRead{Pre,Post} (using gzip)', function()
```

Comment:
if it fits on one line, writing it out like `BufReadPre,BufReadPost` makes grepping easier.